lib/apiservers/engine/backends/archive.go

```diff
@@ -0,0 +1,553 @@
+//				archive header source -		file.txt
+//
+//			The correct writer would be the one corresponding to mount 1.
+//
+//		case 3:
+//				containerDestPath -			/
+//				archive header source -		mnt/a/file.txt
+//
+//			The correct writer would be the one corresponding to mount 1
+//
+// As demonstrated above, the mount prefix and writer cannot be determined with just the
+// container destination path.  It must be combined with the actual asset's name.
+func (wm *ArchiveStreamWriterMap) WriterForAsset(proxy VicContainerProxy, cid, containerDestPath string, assetHeader tar.Header) (io.WriteCloser, error) {
+	defer trace.End(trace.Begin(assetHeader.Name))
+
+	var err error
+	var streamWriter io.WriteCloser
+
+	aw, err := wm.FindArchiveWriter(containerDestPath, assetHeader.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Perform the lazy initialization here.
+	if aw.writer == nil {
+		// lazy initialize.
+		log.Debugf("Lazily initializing import stream for %s", aw.mountPoint.Destination)
+		var deviceID string
+		if aw.mountPoint.Destination == "/" {
+			// Special case. / refers to container VMDK and not a volume vmdk.
+			deviceID = cid
+		} else {
+			deviceID = aw.mountPoint.Name
+		}
+		streamWriter, err = proxy.ArchiveImportWriter(context.Background(), "container", deviceID, aw.filterSpec)
+		if err != nil {
+			err = fmt.Errorf("Unable to initialize import stream writer for mount prefix %s", aw.mountPoint.Destination)
+			log.Errorf(err.Error())
+			return nil, err
+		}
+		aw.writer = streamWriter
+	} else {
+		streamWriter = aw.writer
+	}
+
+	return streamWriter, nil
+}
+
+// Close visits all the archive writer in the trie and closes the actual io.WritCloser
+func (wm *ArchiveStreamWriterMap) Close() {
+	defer trace.End(trace.Begin(""))
+
+	closeStream := func(prefix patricia.Prefix, item patricia.Item) error {
+		if aw, ok := item.(*ArchiveWriter); ok && aw.writer != nil {
+			aw.writer.Close()
+		}
+		return nil
+	}
+
+	wm.prefixTrie.Visit(closeStream)
+}
+
+// FindArchiveReaders finds all archive readers that are within the container source path.  For example,
+//
+//	mount A -			/mnt/A
+//	mount B -			/mnt/B
+//	mount AB -			/mnt/A/AB
+//	base container -	/
+//
+//	container source path - /mnt/A
+//
+// For the above example, this function returns the readers for mount A and mount AB but not the
+// readers for / or mount B.
+func (rm *ArchiveStreamReaderMap) FindArchiveReaders(containerSourcePath string) ([]ArchiveReader, error) {
+	defer trace.End(trace.Begin(containerSourcePath))
+
+	var nodes []ArchiveReader
+	var startingNode ArchiveReader
+	var err error
+
+	findStartingPrefix := func(prefix patricia.Prefix, item patricia.Item) error {
+		if _, ok := item.(ArchiveReader); !ok {
+			return fmt.Errorf("item not ArchiveWriter")
+		}
+
+		startingNode = item.(ArchiveReader)
+		return nil
+	}
+
+	// go function used later for searching
+	walkPrefixSubtree := func(prefix patricia.Prefix, item patricia.Item) error {
```

Comment:
this method looks similar to findStartingPrefix. One suggestion is to combine them, with a switch to differentiate if need to loop all available items.